pkg/api/v1/types.go

```diff
@@ -1428,11 +1428,36 @@ type PreferredSchedulingTerm struct {
 	Preference NodeSelectorTerm `json:"preference" protobuf:"bytes,2,opt,name=preference"`
 }
 
+const (
+	// This annotation key will be used to contain an array of v1 JSON encoded Containers
+	// for init containers. The annotation will be placed into the internal type and cleared.
+	PodInitContainersAnnotationKey = "pod.alpha.kubernetes.io/init-containers"
+	// This annotation key will be used to contain an array of v1 JSON encoded
+	// ContainerStatuses for init containers. The annotation will be placed into the internal
+	// type and cleared.
+	PodInitContainerStatusesAnnotationKey = "pod.alpha.kubernetes.io/init-container-statuses"
+)
+
 // PodSpec is a description of a pod.
 type PodSpec struct {
 	// List of volumes that can be mounted by containers belonging to the pod.
 	// More info: http://releases.k8s.io/HEAD/docs/user-guide/volumes.md
 	Volumes []Volume `json:"volumes,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=volumes"`
+	// List of initialization containers belonging to the pod.
+	// Init containers are executed in order prior to containers being started. If any
+	// init container fails, the pod is considered to have failed and is handled according
+	// to its restartPolicy. The name for an init container or normal container must be
+	// unique among all containers.
+	// Init containers may not have Lifecycle actions, Readiness probes, or Liveness probes.
+	// The resourceRequirements of an init container are taken into account during scheduling
+	// by finding the highest request/limit for each resource type, and then using the max of
+	// of that value or the sum of the normal containers. Limits are applied to init containers
+	// in a similar fashion.
+	// Init containers cannot currently be added or removed.
+	// Init containers are in alpha state and may change without notice.
+	// Cannot be updated.
```

Comment:
I mostly wanted to avoid having to tear down all subsequent init containers
and all app containers in the event of a change.  Today it only tears down
the container.  I'll leave a follow on in the proposal?

On Wed, May 4, 2016 at 10:23 PM, Prashanth B <notifications@github.com>
wrote:

> In pkg/api/v1/types.go
> <https://github.com/kubernetes/kubernetes/pull/23567#discussion_r62142668>
> :
>
> >  // PodSpec is a description of a pod.
> >  type PodSpec struct {
> >  	// List of volumes that can be mounted by containers belonging to the pod.
> >  	// More info: http://releases.k8s.io/HEAD/docs/user-guide/volumes.md
> >  	Volumes []Volume `json:"volumes,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=volumes"`
> > +	// List of initialization containers belonging to the pod.
> > +	// Init containers are executed in order prior to containers being started. If any
> > +	// init container fails, the pod is considered to have failed and is handled according
> > +	// to its restartPolicy. The name for an init container or normal container must be
> > +	// unique among all containers.
> > +	// Init containers may not have Lifecycle actions, Readiness probes, or Liveness probes.
> > +	// The resourceRequirements of an init container are taken into account during scheduling
> > +	// by finding the highest request/limit for each resource type, and then using the max of
> > +	// of that value or the sum of the normal containers. Limits are applied to init containers
> > +	// in a similar fashion.
> > +	// Init containers cannot currently be added or removed.
> > +	// Init containers are in alpha state and may change without notice.
> > +	// Cannot be updated.
>
> So they're not like "normal" containers because I can't update the image
> on a crashlooping container, right? I guess the comment say immutable, but
> I still expected the same image update caveat.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23567/files/fabc8f18ac2996f19fe403b2be7d42bf0f8f8849#r62142668>
>